### PR TITLE
Add varlen support to AOTriton's Flash Attention

### PR DESCRIPTION
**Pull request by xinyazhang**
Varlen Flash Attention is implemented by two new APIs: `attn_fwd_compact_varlen` and `attn_bwd_compact_varlen`, with the same set of kernels. Check `include/aotriton/flash.h` for their details.

Note:
1. the bias tensor input `b` is reserved. Uses should pass `TensorView<4>::get_null_tensor()` for this argument for now. Any other inputs are not supported nor tested.
2. The varlen API still expects Rank-4 tensors, which uniforms the code b/w varlen and non-varlen. The API also expects the size of the first dimension (batch) is exact 1. Extra batches will not be processed. This interface is slightly different from Tri Dao's implementation (https://github.com/Dao-AILab/flash-attention/blob/320fb59487658f033f56711efd3d61b7c7a6f8f3/csrc/flash_attn/flash_api.cpp#L500-L502), and users are supposed to use `torch.transpose` and `torch.unsqueeze` to match the API.
3. This PR also refactored the Triton kernel, and unified the compiled data type of sequence length related arguments to int32_t

### CODE REVIEW

**Review by wangye805 — COMMENTED** [review]
I took a quick look and realized your implementation should work for the qkv_layout of THD in NVTE. 

How about for the layout of BSHD/BHSD with padding mask and cu_seqlen, as we discussed before?

**Comment by xinyazhang** [comment]
> How about for the layout of BSHD/BHSD with padding mask and cu_seqlen, as we discussed before?

It's in the triton kernel but not exposed as C++ API. They will be added in a separate PR as `attn_?wd_padded_varlen`

**Comment by wangye805** [comment]
> > How about for the layout of BSHD/BHSD with padding mask and cu_seqlen, as we discussed before?
> 
> It's in the triton kernel but not exposed as C++ API. They will be added in a separate PR as `attn_?wd_padded_varlen`

Do you mind pointing me to the triton source codes for padded_varlen?

**Comment by xinyazhang** [comment]
> Do you mind pointing me to the triton source codes for padded_varlen?

https://github.com/ROCm/aotriton/blob/c8551b171b8842b390a6fc03c217fb2101820d72/tritonsrc/fwd_kernel.py#L73-L85

**Review by wangye805 — APPROVED** [review]
LGTM

**Review by groenenboomj — APPROVED** [review]
Looking to add more support for more layouts later but that will be a different piece of work.  This looks to meet needs for varlen.

File: include/aotriton/flash.h [hunk 1]
```diff
@@ -31,6 +31,25 @@ attn_fwd(T4 q, // batch_size x num_heads x seqlen_q x head_size
          bool is_causal,
          aotriton::Stream stream);
 
+hipError_t
+attn_fwd_compact_varlen(T4 q, // 1 x num_heads x total_q x head_size, total_q := \sum_{i=0}^{b} s_i
```
[the remark this answers is not in the document]
Comment (xinyazhang) [reply include/aotriton/flash.h:35]: What do you mean by layout flag? If it's about performance for packed QKV I'm currently thinking of classification methods without using user supplied flags.

**Pull request merged**